pkg/controller/atlascustomrole/custom_role.go

```diff
@@ -0,0 +1,106 @@
+package atlascustomrole
+
+import (
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/internal/translation/customroles"
+	akov2 "github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/controller/workflow"
+)
+
+type roleController struct {
+	ctx       *workflow.Context
+	projectID string
+	service   customroles.CustomRoleService
+	role      *akov2.AtlasCustomRole
+	dpEnabled bool
+}
+
+func handleCustomRole(ctx *workflow.Context, akoCustomRole *akov2.AtlasCustomRole, dpEnabled bool) workflow.Result {
+	ctx.Log.Debug("starting custom role processing")
+	defer ctx.Log.Debug("finished custom role processing")
+
+	r := roleController{
+		ctx:       ctx,
+		service:   customroles.NewCustomRoles(ctx.SdkClient.CustomDatabaseRolesApi),
+		role:      akoCustomRole,
+		dpEnabled: dpEnabled,
+	}
+
+	return r.Reconcile()
+}
+
+func (r *roleController) Reconcile() workflow.Result {
+	r.projectID = r.role.Spec.ProjectIDRef.ID
+
+	currentCustomRoles, err := r.service.List(r.ctx.Context, r.projectID)
+	if err != nil {
+		return workflow.Terminate(workflow.ProjectCustomRolesReady, err.Error())
+	}
+
+	roleFoundInAtlas := false
+	roleDeleted := !r.role.GetDeletionTimestamp().IsZero()
+
+	roleInAKO := customroles.NewCustomRole(&r.role.Spec.Role)
+	var roleInAtlas customroles.CustomRole
+	for _, role := range currentCustomRoles {
+		if role.Name == roleInAKO.Name {
+			roleFoundInAtlas = true
+			roleInAtlas = role
+			break
+		}
+	}
+
+	switch {
+	case !roleFoundInAtlas && !roleDeleted:
+		return r.create(roleInAKO)
+	case roleFoundInAtlas && !roleDeleted:
+		return r.update(roleInAKO, roleInAtlas)
+	case roleFoundInAtlas && roleDeleted && !r.dpEnabled:
+		return r.delete(roleInAtlas)
+	}
+
+	return r.idle()
+}
+
+func (r *roleController) create(role customroles.CustomRole) workflow.Result {
+	err := r.service.Create(r.ctx.Context, r.projectID, role)
+	if err != nil {
+		return r.terminate(workflow.AtlasCustomRoleNotCreated, err)
+	}
+	return r.idle()
+}
+
+func (r *roleController) update(roleInAKO, roleInAtlas customroles.CustomRole) workflow.Result {
+	if cmp.Diff(roleInAKO, roleInAtlas, cmpopts.EquateEmpty()) == "" {
```

Comment:
We still use it in custom roles. Also, the panics are mitigated by `cmpopts.IgnoreUnexported()`

Reply:
This is not only about panics, as per go doc, see https://pkg.go.dev/github.com/google/go-cmp/cmp:
>  It is intended to only be used in tests, as performance is not a goal and it may panic if it cannot compare the values. Its propensity towards panicking means that its unsuitable for production environments where a spurious panic may be fatal.

I highly discourage us from using it in new production code. While `reflect.DeepEquals` is not perfect and has pitfalls, we want to implement custom semantic `Equals` statements in the translation layer if necessary, cc @josvazg .

Reply:
Yes, please, no more `cmp.Diff` calls on new or touched code.